Fix OCIObjectStore save_overwrite=False bug

eracah opened the pull request:
# What does this PR do?

## Background of the bug.
When saving a new object to `OCIObjectStore` with `save_overwrite=False`, the worker tasked with uploading the object will first check if the object is already in OCI first (by calling the `get_object` function of the oci sdk) before uploading it. The way it confirms that the object is not there is by catching a `FileNotFoundError`. The bug exists because the `OCIObjectStore` throws an `oci.exceptions.ServiceError` that goes uncaught instead of a .`FileNotFoundError`.

## Fix
We wrap all `oci.object_storage.ObjectStorageClient.get_object` calls in a try except block and reraise any `oci.exceptions.ServiceError` as `FileNotFoundErrors`. fixes CO-1874 and CO-1903

### Along for the ride
* Added the ability for RemoteUploaderDownloader workers to send their exception to the parent, so the user gets more informative errors instead of just: `'Upload worker crashed. Please check the logs.'` This fixes CO-1467

* Fixed a bug (CO-1811) where one of the workers has an error, but the last batch has already happened, so RemoteUploaderDownloader doesn't know that a worker has failed, but meanwhile there are still tasks in the queue, so the `wait_for_workers` function just runs forever. The fix is to check if the exceptions_queue is empty before entering the busy wait for workers to finish loop. Fixes CO-1811

## Tests
### Manual Tests (local and remote with 2 gpu's)

- [x] Save with Non-existent bucket: **errors with proper error.**
- [x] Save with nonexistent object and save_overwrite=False: **works.**
- [x] Save with nonexistent object  with save_overwrite=True: **works.**
- [x] Save with existing object save_overwrite=False: **errors with proper error.**
- for short training runs on gpu's this triggers the bug from CO-1811 if this fix is not in place.
- will error with a very vague error due to CO-1467 if this fix is not in place.
- [x] Save with existing object save_overwrite=True: **works.**
- [x] Load with nonexistent object_name: **errors with proper error.**
- [x] Load with existent path: **works.**

### Autoresume Manual Tests
```
    trainer = Trainer(
        model=mnist_model(num_classes=10),
        train_dataloader=train_dataloader,
        max_duration="8ba", 
        save_interval="4ba",
        save_folder='oci://evan-test/auto_resume_test/{run_name}',
        save_overwrite=False,
        save_num_checkpoints_to_keep=0,
        log_to_console=True,
        console_log_interval='1ba',
        autoresume=True,
        run_name='evans-fun-run',   
    )
    trainer.fit()

    trainer = Trainer(
        model=mnist_model(num_classes=10),
        train_dataloader=train_dataloader,
        max_duration="20ba", 
        save_interval="4ba",
        save_folder='oci://evan-test/auto_resume_test/{run_name}',
        save_overwrite=False,
        save_num_checkpoints_to_keep=0,
        log_to_console=True,
        console_log_interval='1ba',
        autoresume=True,
        run_name='evans-fun-run',   
    )
    trainer.fit()
```
- [x] local autoresume (script above): **works**
- [x] remote 2 gpu autoresume (same script as above basically): **works**
   

### Manual Tests (training run)
- [x] Run examples/llm training and save checkpoints to OCI with save_overwrite: False: **works!!!**
- [x] Run examples/llm training and save checkpoints to OCI with save_overwrite: False pointed at the same folder: **errors as it should: `FileExistsError: Object oci://<foo>.pt already exists, but allow_overwrite was set to False.`

### Unit Tests
- [x] Unit test for RemoteUploaderDownloader exceptions queue
- [x] Unit tests for new exception handling in OCIObjectStore


# What issue(s) does this change relate to?
fixes CO-1874
fixes CO-1467
fixes CO-1811
fixes CO-1903

mvpatel2000 commented on composer/utils/object_store/oci_object_store.py:
```diff
@@ -16,6 +16,26 @@
+    # If it's an oci service error with code: ObjectNotFound or status 404
+    if isinstance(e, oci.exceptions.ServiceError):
+        if e.status == 404: # type: ignore
+            if e.code == 'ObjectNotFound':  # type: ignore
+                raise FileNotFoundError(f'Object {uri} not found. {e.message}') from e
```
```suggestion
            if e.code == 'ObjectNotFound':  # type: ignore
                raise FileNotFoundError(f'Object {uri} not found.') from e
```
The message should be shown because you're doing from e

eracah replied:
Actually, for oci.exceptions.service_errors it's a lot prettier to extract the message because if you raise the error it prints a lot of other stuff. See [this code](https://github.com/oracle/oci-python-sdk/blob/master/src/oci/exceptions.py#L64). They actually pass the error_details to the Exception constructor not the message, so it's a lot more verbose if you raise the error without printing the message first

mvpatel2000 commented on composer/utils/object_store/oci_object_store.py:
```diff
@@ -16,6 +16,26 @@
+        if e.status == 404: # type: ignore
+            if e.code == 'ObjectNotFound':  # type: ignore
+                raise FileNotFoundError(f'Object {uri} not found. {e.message}') from e
+            if e.code == 'BucketNotFound':
+                raise ValueError(f'Bucket specified in {uri} not found. {e.message}') from e
```
```suggestion
            if e.code == 'BucketNotFound':
                raise ValueError(f'Bucket specified in {uri} not found.') from e
```
as above

eracah replied:
same as above

mvpatel2000 commented on composer/utils/object_store/oci_object_store.py:
```diff
@@ -16,6 +16,26 @@
 
 
+def _reraise_oci_errors(uri: str, e: Exception):
+    import oci
```
does this need import gating

eracah replied:
no because it's only called in functions called after the `__init__` is called. But I could add it anyway

eracah replied:
added it

mvpatel2000 commented on composer/utils/object_store/oci_object_store.py:
```diff
@@ -42,29 +62,34 @@ def __init__(
+        except Exception as e:
+            _reraise_oci_errors(self.get_uri(object_name), e)
+
+        return int(response.data.headers['Content-Length'])
```
missing response status 200 check? is that no longer necessary

eracah replied:
I could add it back in

eracah replied:
done

mvpatel2000 commented on composer/loggers/remote_uploader_downloader.py:
```diff
@@ -450,14 +457,18 @@ def wait_for_workers(self):
         during this time will error.
         """
         # Verify enqueue thread has processed all tasks
-        while True:
+        # But make sure to only enter this loop if a worker somewhere else hasn't thrown a fatal exception.
```
Suggestion: # Verify enqueue thread has processed all tasks unless a worker threw an exception

eracah replied:
fixed

mvpatel2000 commented on composer/loggers/remote_uploader_downloader.py:
```diff
@@ -450,14 +457,18 @@ def wait_for_workers(self):
             time.sleep(0.2)  # Yield lock for enqueue thread
-        # Verify all tasks have been completed
-        while not self._file_upload_queue.empty():
+        # Verify all tasks have been completed unless there are exceptions that have been thrown.
```
```suggestion
        # Verify all tasks have been completed unless a worker threw an exception
```

mvpatel2000 reviewed (commented):
Can we get a unit test for the exception stuff?

dakinggg reviewed (commented):
Could you please also test autoresume? Both the initial run (with autoresume=True), and the subsequent resumption.

eracah commented:
> Can we get a unit test for the exception stuff?

Done

eracah commented:
> Could you please also test autoresume? Both the initial run (with autoresume=True), and the subsequent resumption.

Done!

eracah commented:
Ok @mvpatel2000, @dakinggg, I addressed the comments and added unit tests for:
* RemoteUploaderDownloader exception queue
* OCIObjectStore

and manual tests for:
* autoresume local and remote
* LLM run with checkpoints existing

please give it another look-see!

mvpatel2000 commented on tests/loggers/test_remote_uploader_downloader.py:
```diff
@@ -287,3 +293,45 @@ def test_valid_backend_names():
         _ = remote_uploader_downloader.remote_backend
+
+
+@pytest.mark.filterwarnings(r'ignore:Exception in thread:pytest.PytestUnhandledThreadExceptionWarning')
```
is this necessary

eracah replied:
yes. I added a comment explaining

dakinggg commented on tests/loggers/test_remote_uploader_downloader.py:
```diff
@@ -119,15 +119,21 @@ def object_store_test_helper(
+                            f'Object local://{remote_file_name} already exists, but allow_overwrite was set to False.'):
                         remote_uploader_downloader.run_event(Event.BATCH_END, dummy_state, logger)
 
-                    with pytest.raises(RuntimeError):
```
this is no longer an error?

eracah replied:
it is, but the BATCH_END call would normally exit the process. It's only a technicality that pytest catches it and technically we could run another event for EPOCH_END. I updated it to separately test BATCH_END and EPOCH_END with a parametrization

dakinggg commented on tests/utils/object_store/test_oci_object_store.py:
```diff
@@ -34,59 +41,135 @@ def test_oci_obj_store(mock_bucket_name, monkeypatch):
 
-def test_upload_object(test_oci_obj_store, monkeypatch, tmp_path, mock_bucket_name):
-    pytest.importorskip('oci')
+@pytest.mark.skipif(not _OCI_INSTALLED, reason='You must have oci installed to run this test')
```
we usually do this by just `oci = pytest.importorskip(oci)` in the test itself

eracah replied:
ok switched it back
